### PR TITLE
Adding VQGAN, interface to pretrained model, dataset class + download script

### PR DESCRIPTION
**Pull request by RamiMatar**

### CODE REVIEW

File: data.py [hunk 1]
```diff
@@ -70,12 +74,12 @@ def get_dataset(path):
 
 # TODO: split this data loader into 85% train, 5% validation, 10% test
```
Comment (Maxusmusti) [inline data.py]: Is this still TODO or leftover comment?

File: download_hf.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+download_hf.py
+
+This script downloads images and their metadata from a Hugging Face dataset using the specified parameters.
+It supports datasets with image bytes or URLs and saves the images and metadata in the specified output directory.
+It uses parallel workers for faster downloading and processes the dataset without loading all the data at once.
+
+Usage:
+    python download_hf.py --dataset_name <dataset_name> [--output_dir <output_dir>] [--rows <rows>] [--workers <workers>] [--image_bytes_key <image_bytes_key>] [--url_key <url_key>] [--metadata_columns <metadata_columns>...]
+
+Arguments:
```
Comment (Maxusmusti) [inline download_hf.py:22]: Actually very nice docstring

File: vqgan/decoder.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        # end
+        self.norm_out = Normalize(block_in)
+        self.conv_out = torch.nn.Conv2d(block_in,
+                                        out_ch,
+                                        kernel_size=3,
+                                        stride=1,
+                                        padding=1)
+
+    def forward(self, z):
+        #assert z.shape[1:] == self.z_shape[1:]
```
Comment (Maxusmusti) [inline vqgan/decoder.py:80]: Should we remove "dead" comments?

File: vqgan/decoder.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        # end
+        if self.give_pre_end:
+            return h
+
+        h = self.norm_out(h)
+        h = nonlinearity(h)
+        h = self.conv_out(h)
+        if self.tanh_out:
+            h = torch.tanh(h)
+        return h
```
Comment (Maxusmusti) [inline vqgan/decoder.py]: Missing newline EOF

File: vqgan/discriminator.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+            nn.LeakyReLU(0.2, True)
+        ]
+
+        sequence += [
+            nn.Conv2d(ndf * nf_mult, 1, kernel_size=kw, stride=1, padding=padw)]  # output 1 channel prediction map
+        self.main = nn.Sequential(*sequence)
+        print(self.main)
+    def forward(self, input):
+        """Standard forward."""
+        return self.main(input)
```
Comment (Maxusmusti) [inline vqgan/discriminator.py]: missing newline at eof

File: vqgan/interface.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        dec = self.decoder(quant)
+        return dec
+    
+
+def pretrained_vqgan():
+    args = get_default_args()
+    ckpt_path = default_args.ckpt_path
+    model = VQInterface(args)
+    model.init_from_ckpt(ckpt_path)
+    return model
```
Comment (Maxusmusti) [inline vqgan/interface.py]: missing newline

File: vqgan/encoder.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        h = self.mid.block_1(h, temb)
+        h = self.mid.attn_1(h)
+        h = self.mid.block_2(h, temb)
+
+        # end
+        h = self.norm_out(h)
+        h = nonlinearity(h)
+        h = self.conv_out(h)
+        return h
+
```
Comment (Maxusmusti) [inline vqgan/encoder.py]: extra newline

File: vqgan/loss.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+                logits_fake = self.discriminator(torch.cat((reconstructions.contiguous().detach(), cond), dim=1))
+
+            disc_factor = adopt_weight(self.disc_factor, global_step, threshold=self.discriminator_iter_start)
+            d_loss = disc_factor * self.disc_loss(logits_real, logits_fake)
+
+            log = {"{}/disc_loss".format(split): d_loss.clone().detach().mean(),
+                   "{}/logits_real".format(split): logits_real.detach().mean(),
+                   "{}/logits_fake".format(split): logits_fake.detach().mean()
+                   }
+            return d_loss, log
```
Comment (Maxusmusti) [inline vqgan/loss.py]: missing newline

File: vqgan/lpips.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+        return out
+
+
+def normalize_tensor(x,eps=1e-10):
+    norm_factor = torch.sqrt(torch.sum(x**2,dim=1,keepdim=True))
+    return x/(norm_factor+eps)
+
+
+def spatial_average(x, keepdim=True):
+    return x.mean([2,3],keepdim=keepdim)
```
Comment (Maxusmusti) [inline vqgan/lpips.py]: missing newline

**Review by Maxusmusti — CHANGES_REQUESTED** [review]
LGTM, just some minor nits

File: data.py (hunk 1 above)
Comment (RamiMatar) [reply data.py]: we can remove this

File: vqgan/decoder.py (hunk 3 above)
Comment (RamiMatar) [reply vqgan/decoder.py:80]: keep for now, some commented lines can be useful for training later

File: data.py (hunk 1 above)
Comment (Maxusmusti) [inline data.py]:
```suggestion

```

**Review by Maxusmusti — APPROVED** [review]
LGTM
